src/libmongoc/doc/mongoc_uri_t.rst

```diff
@@ -106,6 +106,13 @@ MONGOC_URI_SRVMAXHOSTS                     srvmaxhosts                       0
 
 Setting any of the \*timeoutMS options above to ``0`` will be interpreted as "use the default value".
 
+.. warning::
+
+  Setting any of the \*timeoutMS options above to either ``0`` or a negative value is discouraged due to unspecified and inconsistent behavior.
+  The "default value" historically specified as a fallback for ``0`` or a negative value is NOT related to the default values for the \*timeoutMS options documented above.
```

Comment:
> is NOT related to the default values for the \*timeoutMS options documented above.

This is due to the observation that the "default value" set given a negative timeout [appears to be](https://github.com/mongodb/mongo-c-driver/blob/fc1cba77b671f5c8c942d450a20c7c306fe909d4/src/libmongoc/src/mongoc/mongoc-stream.c#L160) with the undocumented `MONGOC_DEFAULT_TIMEOUT_MSEC` constant which appears to have no direct relation to any of the documented URI default values. Furthermore, this is specific to "write" paths; it is unclear to me how a negative value is interpreted in the "read" paths, which have [no equivalent check](https://github.com/mongodb/mongo-c-driver/blob/fc1cba77b671f5c8c942d450a20c7c306fe909d4/src/libmongoc/src/mongoc/mongoc-stream.c#L237) for negative timeouts before invoking `readv()`.

Reply:
I expect the timeout value passed to `writev` needs to be preserved for backwards compatibility.
A custom `mongoc_stream_t` implementation can be set in `mongoc_client_set_stream_initiator`. An implementation may not expect negative values passed in `writev`.

Since negatives are now documented as "unspecified and inconsistent", consider adding a test to ensure the `writev` handling of negatives is kept for backwards compatibility:

```c
typedef struct {
   mongoc_stream_t vtable;
   struct {
      bool set;
      uint32_t value;
   } last_timeout_applied;
} timeout_tracker_stream_t;

static ssize_t
timeout_tracker_stream_writev (mongoc_stream_t *stream,
                               mongoc_iovec_t *iov,
                               size_t iovcnt,
                               int32_t timeout_msec)
{
   timeout_tracker_stream_t *ttstream = (timeout_tracker_stream_t *) stream;

   BSON_UNUSED (iov);
   BSON_UNUSED (iovcnt);
   ttstream->last_timeout_applied.set = true;
   ttstream->last_timeout_applied.value = timeout_msec;
   return 0;
}

void
timeout_tracker_stream_destroy (mongoc_stream_t *stream)
{
   bson_free (stream);
}

static timeout_tracker_stream_t *
timeout_tracker_stream_new (void)
{
   timeout_tracker_stream_t *ttstream;

   ttstream = bson_malloc0 (sizeof *ttstream);
   ttstream->vtable.type = 999;
   ttstream->vtable.writev = timeout_tracker_stream_writev;
   ttstream->vtable.destroy = failing_stream_destroy;

   return ttstream;
}

// Test special handling of negative write timeout in mongoc_stream_t API.
// Negative values in `mongoc_stream_writev` are applied as a private default.
// This behavior is tested for backwards compatibility with existing mongoc_stream_t implementations.
static void
test_stream_negative_write_timeout (void)
{
   char buf[1] = {'a'};
   mongoc_iovec_t iov[1] = {{.iov_base = buf, .iov_len = sizeof (buf)}};
   timeout_tracker_stream_t *ttstream = timeout_tracker_stream_new ();

#define MONGOC_DEFAULT_TIMEOUT_MSEC (60L * 60L * 1000L)

   // writev with -1 timeout applies a private "default".
   ttstream->last_timeout_applied.set = false;
   mongoc_stream_writev ((mongoc_stream_t *) ttstream,
                         iov,
                         sizeof (iov) / sizeof (iov[0]),
                         -1 /* timeout */);
   ASSERT (ttstream->last_timeout_applied.set);
   ASSERT_CMPINT32 (
      ttstream->last_timeout_applied.value, ==, MONGOC_DEFAULT_TIMEOUT_MSEC);

#undef MONGOC_DEFAULT_TIMEOUT_MSEC

   mongoc_stream_destroy ((mongoc_stream_t *) ttstream);
}
```

Reply:
Added a writev test as requested.